### PR TITLE
Add JournalEntry resource and operations

### LINKED ISSUE #27: Add resource support for JournalEntry (5 hrs)
https://developer.intuit.com/app/developer/qbo/docs/api/accounting/all-entities/journalentry

We need to be able to create, update, read, and delete journal entries, including using Searchers for this

Attribute-wise, we need the following
- Id
- Line
- CurrencyRef
- PrivateNote
- TxnDate

### PR DESCRIPTION
**Pull request by jozefvaclavik**
Fix #27

### CODE REVIEW

File: lib/ledger_sync/adaptors/quickbooks_online/adaptor.rb [hunk 1]
```diff
@@ -44,10 +44,10 @@ def initialize(
         end
 
         def find(resource:, id:)
-          url = "#{oauth_base_uri}/#{underscore(resource)}/#{id}"
+          url = "#{oauth_base_uri}/#{resourcify(resource)}/#{id}"
```
Comment (ryanwjackson) [inline lib/ledger_sync/adaptors/quickbooks_online/adaptor.rb]: Why this change instead of using `Util::StringHelpers`?

Comment (jozefvaclavik) [reply lib/ledger_sync/adaptors/quickbooks_online/adaptor.rb]: its just `.tr('_', '')`. Didn't feel to me that it needs to be as Util. Should I move it as utility helper?

**Comment by ryanwjackson** [comment]
I believe we use it other places is why I put it in a Util.


--
Ryan Jackson
ryanwjackson@gmail.com

File: lib/ledger_sync/adaptors/quickbooks_online/adaptor.rb (hunk 1 above)
Comment (mattgmarcus) [reply lib/ledger_sync/adaptors/quickbooks_online/adaptor.rb]: This might be QBO specific though. I'm fine with it like this

File: lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/create.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                },
+                'TxnDate' => resource.transaction_date.to_s, # Format: YYYY-MM-DD
+                'PrivateNote' => resource.memo,
+                'Line' => resource.line_items.map do |line_item|
+                  {
+                    'DetailType' => 'JournalEntryLineDetail',
+                    'JournalEntryLineDetail' => {
+                      'PostingType': Mapping::LINE_ITEM_TYPES[line_item.entry_type],
+                      'AccountRef' => {
+                        'value' => line_item.account&.ledger_id
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/create.rb]: Why make this nil safe?

File: lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/update.rb [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                },
+                'TxnDate' => resource.transaction_date.to_s, # Format: YYYY-MM-DD
+                'PrivateNote' => resource.memo,
+                'Line' => resource.line_items.map do |line_item|
+                  {
+                    'DetailType' => 'JournalEntryLineDetail',
+                    'JournalEntryLineDetail' => {
+                      'PostingType': Mapping::LINE_ITEM_TYPES[line_item.entry_type],
+                      'AccountRef' => {
+                        'value' => line_item.account&.ledger_id
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/update.rb]: Same Q about nil safe

File: lib/ledger_sync/adaptors/quickbooks_online/journal_entry/searcher.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            @resources ||= begin
+              adaptor
+                .query(
+                  resource: 'journal_entry',
+                  limit: limit,
+                  offset: offset
+                )
+                .map do |c|
+                  LedgerSync::JournalEntry.new(
+                    ledger_id: c.fetch('Id')
```
Comment (mattgmarcus) [inline lib/ledger_sync/adaptors/quickbooks_online/journal_entry/searcher.rb]: Can we pull in the other attributes while you're here?

File: lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/create.rb (hunk 2 above)
Comment (jozefvaclavik) [reply lib/ledger_sync/adaptors/quickbooks_online/journal_entry/operations/create.rb]: In case we don't have account, right now we will let QBO fail, instead of crashing in the lib. This will change once we handle serialisation and dirty.

Right now we would send `nil` to QBO, which could remove account if its optional attribute. With serialisation we will not serialise it if we don't provide account.

File: lib/ledger_sync/adaptors/quickbooks_online/journal_entry/searcher.rb (hunk 4 above)
Comment (jozefvaclavik) [reply lib/ledger_sync/adaptors/quickbooks_online/journal_entry/searcher.rb]: Yes, but this will get automated once we do serialisation.
